libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1051 @@
+          *       // add dan account to table           - see put example
+          *
+          *       auto dan = addresses.index.full_name.find("Dan Larimer");
+          *       eosio::check(dan != addresses.index.full_name.end(), "Dan Larimer is not in the table");
+          *       eosio::check(dan.value().city == "Blacksburg", "Got the wrong value");
+          *    }
+          * @endcode
+          */
+         T value() const {
+            using namespace detail;
+
+            eosio::check(itr_stat != kv_it_stat::iterator_end, "Cannot read end iterator");
+            eosio::check(data_size > 0, "Cannot read a value of size 0");
+
+            uint32_t actual_value_size;
+            uint32_t offset = 0;
+
+            void* buffer = data_size > detail::max_stack_buffer_size ? malloc(data_size) : alloca(data_size);
+            auto stat = internal_use_do_not_use::kv_it_value(itr, offset, (char*)buffer, data_size, actual_value_size);
+
+            eosio::check(static_cast<kv_it_stat>(stat) != kv_it_stat::iterator_end, "Error reading value");
+
+            if (idx->name == idx->tbl->primary_index->name) {
+               datastream<const char*> ds((char*)buffer, actual_value_size);
+
+               T val;
+               ds >> val;
+               if (data_size > detail::max_stack_buffer_size) {
+                  free(buffer);
+               }
+               return val;
+            } else {
+               uint32_t actual_data_size;
+               auto success = internal_use_do_not_use::kv_get(db, contract_name.value, (const char*)buffer, actual_value_size, actual_data_size);
+               eosio::check(success, "failure getting primary key");
+
+               void* pk_buffer = actual_data_size > detail::max_stack_buffer_size ? malloc(actual_data_size) : alloca(actual_data_size);
+               auto copy_size = internal_use_do_not_use::kv_get_data(db, 0, (char*)pk_buffer, actual_data_size);
+
+               eosio::check(copy_size > 0, "failure getting primary index data");
+
+               datastream<const char*> ds((char*)pk_buffer, copy_size);
+
+               T val;
+               ds >> val;
+
+               if (actual_data_size > detail::max_stack_buffer_size) {
+                  free(pk_buffer);
+               }
+               if (data_size > detail::max_stack_buffer_size) {
+                  free(buffer);
+               }
+               return val;
+            }
+         }
+
+         iterator& operator++() {
+            eosio::check(itr_stat != kv_it_stat::iterator_end, "cannot increment end iterator");
+            itr_stat = static_cast<kv_it_stat>(internal_use_do_not_use::kv_it_next(itr));
+            return *this;
+         }
+
+         iterator& operator--() {
+            itr_stat = static_cast<kv_it_stat>(internal_use_do_not_use::kv_it_prev(itr));
+            eosio::check(itr_stat != kv_it_stat::iterator_end, "incremented past the beginning");
+            return *this;
+         }
```

Comment:
You should also provide `operator++(int)` and `operator--(int)`

Reply:
Bucky and I decided intentionally not to, so that people don't inadvertently use them and suffer the copy penalty.

Reply:
Plus the intrinsics don't support copying iterators